### DISCUSSION TITLE
cbc:DurationMeasure

### DISCUSSION THREAD

**Discussion by rengit62** [OP-TED/eForms-SDK#421] [Q&A] [answered]
Hi Lab
in example notices and in doc I found 3 samples of cbc:DurationMeasure

Tender Validity Deadline (BT-98)
<cac:TenderValidityPeriod>
    <cbc:DurationMeasure unitCode="MONTH">4</cbc:DurationMeasure>
</cac:TenderValidityPeriod>

Procedure planned period
<cac:PlannedPeriod>
    <cbc:DurationMeasure unitCode="DAY">3</cbc:DurationMeasure>
</cac:PlannedPeriod>

<cac:PlannedPeriod>
				<cbc:StartDate>2021-04-23+02:00</cbc:StartDate>
				<cbc:DurationMeasure unitCode="YEAR">4</cbc:DurationMeasure>
			</cac:PlannedPeriod>

How can I choose among YEAR, MONTH and DAY ?
Thanks
Renato

**Comment by rouschr** [comment]
Can you maybe rephrase this question? I am not sure what the problem is.

**Reply by rengit62** [reply]
OK
from fields.json  : 
 **"id" : "BT-36-Lot"**
***********************
"xpathRelative" : "cbc:DurationMeasure",

xsd : 
 <xsd:complexType name="DurationMeasureType">
      <xsd:simpleContent>
         <xsd:restriction base="udt:MeasureType"/>
      </xsd:simpleContent>
   </xsd:complexType>
  
 Measure. Unit. Code           as  (MeasureType)/@unitCode

 <ccts:Definition>A numeric value determined by measuring an object using a specified unit of measure.</ccts:Definition>
       
Which is this unit of measure ? 
As I wrote, in notice examples I found DAY, MONTH and YEAR. How should I choose among these ?
In the doc : 
The "cbc:DurationMeasure" element shall be used with the specification of the unit **(e.g. "days")** and its total number to cover the whole period. Unit of measure shall correspond to one of the duration units defined in the [timeperiod](https://op.europa.eu/en/web/eu-vocabularies/at-dataset/-/resource/dataset/timeperiod) codelist.

<cac:PlannedPeriod>
    <cbc:DurationMeasure unitCode="DAY">3</cbc:DurationMeasure>
</cac:PlannedPeriod>



Thanks and Regards
Renato

**Reply by rouschr** [reply]
Maybe @YvesJo can help as this is more business oriented

**Reply by YvesJo** [reply]
The unit of measure should be one of the tailored list [duration-unit](https://github.com/OP-TED/eForms-SDK/blob/develop/codelists/timeperiod_duration-unit.gc), subset of timeperiod.

**Reply by rengit62** [reply]
OK Yves
in front-end form there is the field Period-duration ,  in duration-unit there are 4 rows , YEAR, MONTH, WEEK and DAY.
The number set in Period-duration field refers to YEAR, MONTH, WEEK or DAY ?
![Duration](https://user-images.githubusercontent.com/52211037/235681277-a165b4fa-3062-4237-b85a-71db67ee0e82.jpg)
DAY ?

**Reply by YvesJo** [reply]
I guess this field is for the value and this UI is missing the additional implicit field for the unit.

**Reply by rouschr** [reply, marked as answer]
In the SDK notice-types/29.json you can see that we have this to define the UI:
```
 }, {
          "id" : "GR-Lot-PlannedDuration-Period",
          "contentType" : "group",
          "displayType" : "GROUP",
          "description" : "Duration Period",
          "_label" : "group|name|GR-Lot-PlannedDuration-Period",
          "content" : [ {
            "id" : "BT-36-Lot",
            "contentType" : "field",
            "displayType" : "TEXTBOX",
            "description" : "Duration Period",
            "_label" : "field|name|BT-36-Lot"
          } ]
        }, {
          "id" : "BT-538-Lot",
          "contentType" : "field",
          "displayType" : "COMBOBOX",
          "description" : "Duration Other",
          "_label" : "field|name|BT-538-Lot"
        } ]
      }
```
I think that because it is so called "implicit" we have no field for it and if we have no field there is also nothing in the NTD (notice type definition).
We do have a codelist called "duration-unit" (see timeperiod_duration-unit.gc) but I do not see it in the fields.json like I see "measurement-unit".
This looks like something we must add.

**Reply by elerik** [reply]
will this be implemented in version 1.8 ? 
in 1.7 its still missing (probably in every notice subtype where are these BT-36-Lot fields)

**Reply by rouschr** [reply]
Sadly not in 1.8 and also not foreseen for 1.9 yet.
But in 1.8 there will be many fixes and improvements in the notice types json files which should help with generation of XML from the UI data.

**Reply by elerik** [reply]
Then i have a question. How can any company go live without this and unknown missing stuff ? unknown because this was find out by accident. 
In other words. There is no guarantee that the validation will take place and that the notice will be submitted. Because in this case validation will be false if customer choose this variation. So plans to going live is technically postponed.

**Reply by jaggaer-gbartucc** [reply]
Hi, we noticed that the issue has not been resolved yet for SDK 1.13. In the notice-types definition, the measurement unit field for the BT-36-Lot field is missing. Could you please tell us when the issue will be resolved? Thanks

**Reply by rouschr** [reply]
Since 1.9 the link to the other field is indicated via **attributes** in the fields.json:

```
{
    "id" : "BT-36-Lot",
    "parentNodeId" : "ND-LotDuration",
    "name" : "Duration Period",
    "btId" : "BT-36",
    "xpathAbsolute" : "/*/cac:ProcurementProjectLot[cbc:ID/@schemeName='Lot']/cac:ProcurementProject/cac:PlannedPeriod/cbc:DurationMeasure",
    "xpathRelative" : "cbc:DurationMeasure",
    "xsdSequenceOrder" : [ { "cbc:DurationMeasure" : 6 } ],
    "type" : "measure",
    "attributes" : [ "BT-36-Lot-Unit" ], <----------------------------- The related field
```

You can dynamically add an input in the form based on "BT-36-Lot-Unit".
Later the XML generation also needs special logic to handle attributes as they are stored in a special location in the XML compared to other values.
There is documentation about it here, search for "attributes":
https://docs.ted.europa.eu/eforms/latest/fields/index.html
